File: example/wgs_conversions_client.py
#!/usr/bin/env python
"""
An example client for using the wgs_conversions services

Dan Pierce
2017-03-13
"""
import sys
import rospy
from wgs_conversions.srv import WgsConversion

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from XYZ to LLA
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
x = 422143.588316416
y = -5361864.950247487
z = 3417050.753609172

print "\nConverting position"
print "\tX: ",x,"\tY: ",y,"\tZ: ",z

rospy.wait_for_service('xyz2lla')

xyz2lla = rospy.ServiceProxy('xyz2lla', WgsConversion)

rsp = xyz2lla(xyz=(x,y,z))
(lat,lon,alt)=rsp.lla

print "\n\tSuccessfully converted from ecef to lla"
print "\t\tLat: ",lat,"\tLong: ",lon,"\tAlt: ",alt

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from LLA to XYZ
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
rospy.wait_for_service('lla2xyz')

lla2xyz = rospy.ServiceProxy('lla2xyz', WgsConversion)

rsp = lla2xyz(lla=(lat,lon,alt))
(x,y,z)=rsp.xyz

print "\n\tSuccessfully converted from lla back to ecef"
print "\t\tX: ",x,"\tY: ",y,"\tZ: ",z

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from LLA to ENU
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
ref_lat = lat + 0.01
ref_lon = lon + 0.01
ref_alt = alt + 10.0

rospy.wait_for_service('lla2enu')

lla2enu = rospy.ServiceProxy('lla2enu', WgsConversion)

rsp = lla2enu(lla=(lat,lon,alt),ref_lla=(ref_lat,ref_lon,ref_alt))
(e,n,u)=rsp.enu

print "\n\tSuccessfully converted from lla to enu"
print "\t\tE: ",e,"\tN: ",n,"\tU: ",u

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from ENU to LLA
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
rospy.wait_for_service('enu2lla')

enu2lla = rospy.ServiceProxy('enu2lla', WgsConversion)

rsp = enu2lla(enu=(e,n,u),ref_lla=(ref_lat,ref_lon,ref_alt))
(lat,lon,alt)=rsp.lla

print "\n\tSuccessfully converted from enu back to lla"
print "\t\tLat: ",lat,"\tLong: ",lon,"\tAlt: ",alt

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from XYZ to ENU
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
rospy.wait_for_service('xyz2enu')

xyz2enu = rospy.ServiceProxy('xyz2enu', WgsConversion)

rsp = xyz2enu(xyz=(x,y,z),ref_lla=(ref_lat,ref_lon,ref_alt))
(e,n,u)=rsp.enu

print "\n\tSuccessfully converted from xyz to enu"
print "\t\tE: ",e,"\tN: ",n,"\tU: ",u

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert from ENU to XYZ
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
rospy.wait_for_service('enu2xyz')

enu2xyz = rospy.ServiceProxy('enu2xyz', WgsConversion)

rsp = enu2xyz(enu=(e,n,u),ref_lla=(ref_lat,ref_lon,ref_alt))
(x,y,z)=rsp.xyz

print "\n\tSuccessfully converted from enu back to xyz"
print "\t\tX: ",x,"\tY: ",y,"\tZ: ",z

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert velocities from XYZ to ENU
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
vx = 23.3883
vy = 2.1984
vz = 0.4462

print "\nConverting velocity"
print "\tvX: ",vx,"\tvY: ",vy,"\tvZ: ",vz

rospy.wait_for_service('xyz2enu_vel')

xyz2enu_vel = rospy.ServiceProxy('xyz2enu_vel', WgsConversion)

rsp = xyz2enu_vel(xyz=(vx,vy,vz),ref_lla=(ref_lat,ref_lon,ref_alt))
(ve,vn,vu)=rsp.enu

print "\n\tSuccessfully converted velocities from xyz to enu"
print "\t\tvE: ",ve,"\tvN: ",vn,"\tvU: ",vu

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert velocities from ENU to XYZ
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
rospy.wait_for_service('enu2xyz_vel')

enu2xyz_vel = rospy.ServiceProxy('enu2xyz_vel', WgsConversion)

rsp = enu2xyz_vel(enu=(ve,vn,vu),ref_lla=(ref_lat,ref_lon,ref_alt))
(vx,vy,vz)=rsp.xyz

print "\n\tSuccessfully converted velocities from enu back to xyz"
print "\t\tvX: ",vx,"\tvY: ",vy,"\tvZ: ",vz

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert covariance from XYZ to ENU
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
Pxyz =  ( (0.1527,-0.0244,0.0267),
	      (-0.0244,0.5772,-0.2432),
	      (0.0267,-0.2432,0.2687) )

print "\nConverting covariance"
print "\tPxyz:\t",Pxyz[0][0],"  ",Pxyz[0][1],"  ",Pxyz[0][2]
print "\t     \t",Pxyz[1][0],"  ",Pxyz[1][1],"  ",Pxyz[1][2]
print "\t     \t",Pxyz[2][0],"  ",Pxyz[2][1],"  ",Pxyz[2][2]

pxyz = []
for i in range(3):
	for j in range(3):
		pxyz.append(Pxyz[i][j])

rospy.wait_for_service('xyz2enu_cov')

xyz2enu_cov = rospy.ServiceProxy('xyz2enu_cov', WgsConversion)

rsp = xyz2enu_cov(xyz_cov=pxyz,ref_lla=(ref_lat,ref_lon,ref_alt))

Penu = ( (rsp.enu_cov[0],rsp.enu_cov[1],rsp.enu_cov[2]),
	     (rsp.enu_cov[3],rsp.enu_cov[4],rsp.enu_cov[5]),
	     (rsp.enu_cov[6],rsp.enu_cov[7],rsp.enu_cov[8]) )

print "\n\tSuccessfully converted covariance from ecef to enu"
print "\t\tPenu:\t",Penu[0][0],"  ",Penu[0][1],"  ",Penu[0][2]
print "\t\t     \t",Penu[1][0],"  ",Penu[1][1],"  ",Penu[1][2]
print "\t\t     \t",Penu[2][0],"  ",Penu[2][1],"  ",Penu[2][2]

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
"" Convert covariance from ENU to XYZ
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
penu = []
for i in range(3):
	for j in range(3):
		penu.append(Penu[i][j])

rospy.wait_for_service('enu2xyz_cov')

enu2xyz_cov = rospy.ServiceProxy('enu2xyz_cov', WgsConversion)

rsp = enu2xyz_cov(enu_cov=penu,ref_lla=(ref_lat,ref_lon,ref_alt))

Pxyz = ( (rsp.xyz_cov[0],rsp.xyz_cov[1],rsp.xyz_cov[2]),
	     (rsp.xyz_cov[3],rsp.xyz_cov[4],rsp.xyz_cov[5]),
	     (rsp.xyz_cov[6],rsp.xyz_cov[7],rsp.xyz_cov[8]) )

print "\n\tSuccessfully converted covariance from enu back to ecef"
print "\t\tPxyz:\t",Pxyz[0][0],"  ",Pxyz[0][1],"  ",Pxyz[0][2]
print "\t\t     \t",Pxyz[1][0],"  ",Pxyz[1][1],"  ",Pxyz[1][2]
print "\t\t     \t",Pxyz[2][0],"  ",Pxyz[2][1],"  ",Pxyz[2][2]

File: example/wgs_conversions_client.cpp
/*
 * An example client for using the wgs_conversions services 
 *
 * Dan Pierce
 * 2017-03-13
 */
#include <ros/ros.h>
#include "wgs_conversions/WgsConversion.h"

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

int main(int argc, char** argv)
{
  ros::init(argc, argv, "wgs_conversions_client");

  std::cout.precision(15);

  ros::NodeHandle nh;
  ros::ServiceClient client;
  wgs_conversions::WgsConversion srv;

  ////////////////////////////////////////////////////////////////////////////////
  // Convert from XYZ to LLA
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("xyz2lla");

  double xyz[3],lla[3],ref_lla[3],enu[3];

  xyz[0]= 422143.588316416;
  xyz[1]= -5361864.950247487;
  xyz[2]= 3417050.753609172;

  std::cout << "\nConverting position\n\tX: " << xyz[0] << "\tY: " << xyz[1] << "\tZ: " << xyz[2] << std::endl;
  
  srv.request.xyz[0] = xyz[0];
  srv.request.xyz[1] = xyz[1];
  srv.request.xyz[2] = xyz[2];

  if (client.call(srv)){    
    lla[0] = srv.response.lla[0];
    lla[1] = srv.response.lla[1];
    lla[2] = srv.response.lla[2];
    std::cout << "\n\tSuccessfully converted from ecef to lla" << std::endl;
    std::cout << "\t\tLat: " << lla[0] << "\tLong: " << lla[1] << "\tAlt: " << lla[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service xyz2lla");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert from LLA to XYZ
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("lla2xyz");

  srv.request.lla[0] = lla[0];
  srv.request.lla[1] = lla[1];
  srv.request.lla[2] = lla[2];

  if (client.call(srv)){
    xyz[0] = srv.response.xyz[0];
    xyz[1] = srv.response.xyz[1];
    xyz[2] = srv.response.xyz[2];
    std::cout << "\n\tSuccessfully converted from lla back to ecef" << std::endl;
    std::cout << "\t\tX: " << xyz[0] << "\tY: " << xyz[1] << "\tZ: " << xyz[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service lla2xyz");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert from LLA to ENU
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("lla2enu");

  ref_lla[0] = lla[0] + 0.01;
  ref_lla[1] = lla[1] + 0.01;
  ref_lla[2] = lla[2] + 10.0;
  
  srv.request.lla[0] = lla[0];
  srv.request.lla[1] = lla[1];
  srv.request.lla[2] = lla[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    enu[0] = srv.response.enu[0];
    enu[1] = srv.response.enu[1];
    enu[2] = srv.response.enu[2];
    std::cout << "\n\tSuccessfully converted from lla to enu" << std::endl;
    std::cout << "\t\tE: " << enu[0] << "\tN: " << enu[1] << "\tU: " << enu[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service lla2enu");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert from ENU to LLA
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("enu2lla");

  srv.request.enu[0] = enu[0];
  srv.request.enu[1] = enu[1];
  srv.request.enu[2] = enu[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    lla[0] = srv.response.lla[0];
    lla[1] = srv.response.lla[1];
    lla[2] = srv.response.lla[2];
    std::cout << "\n\tSuccessfully converted from enu back to lla" << std::endl;
    std::cout << "\t\tLat: " << lla[0] << "\tLong: " << lla[1] << "\tAlt: " << lla[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service enu2lla");
    return 1;
  }
  
  ////////////////////////////////////////////////////////////////////////////////
  // Convert from XYZ to ENU
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("xyz2enu");

  srv.request.xyz[0] = xyz[0];
  srv.request.xyz[1] = xyz[1];
  srv.request.xyz[2] = xyz[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    enu[0] = srv.response.enu[0];
    enu[1] = srv.response.enu[1];
    enu[2] = srv.response.enu[2];
    std::cout << "\n\tSuccessfully converted from ecef to enu" << std::endl;
    std::cout << "\t\tE: " << enu[0] << "\tN: " << enu[1] << "\tU: " << enu[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service xyz2enu");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert from ENU to XYZ
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("enu2xyz");

  srv.request.enu[0] = enu[0];
  srv.request.enu[1] = enu[1];
  srv.request.enu[2] = enu[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    xyz[0] = srv.response.xyz[0];
    xyz[1] = srv.response.xyz[1];
    xyz[2] = srv.response.xyz[2];
    std::cout << "\n\tSuccessfully converted from enu back to ecef" << std::endl;
    std::cout << "\t\tX: " << xyz[0] << "\tY: " << xyz[1] << "\tZ: " << xyz[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service enu2xyz");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert velocities from XYZ to ENU
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("xyz2enu_vel");

  double xyz_vel[3],enu_vel[3];

  xyz_vel[0] = 23.3883;
  xyz_vel[1] = 2.1984;
  xyz_vel[2] = 0.4462;

  std::cout << "\nConverting velocity\n\tvX: " << xyz_vel[0] << "\tvY: " << xyz_vel[1] << "\tvZ: " << xyz_vel[2] << std::endl;

  srv.request.xyz[0] = xyz_vel[0];
  srv.request.xyz[1] = xyz_vel[1];
  srv.request.xyz[2] = xyz_vel[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    enu_vel[0] = srv.response.enu[0];
    enu_vel[1] = srv.response.enu[1];
    enu_vel[2] = srv.response.enu[2];
    std::cout << "\n\tSuccessfully converted velocities from ecef to enu" << std::endl;
    std::cout << "\t\tvE: " << enu_vel[0] << "\tvN: " << enu_vel[1] << "\tvU: " << enu_vel[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service xyz2enu_vel");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert velocities from ENU to XYZ
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("enu2xyz_vel");

  srv.request.enu[0] = enu_vel[0];
  srv.request.enu[1] = enu_vel[1];
  srv.request.enu[2] = enu_vel[2];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    xyz_vel[0] = srv.response.xyz[0];
    xyz_vel[1] = srv.response.xyz[1];
    xyz_vel[2] = srv.response.xyz[2];
    std::cout << "\n\tSuccessfully converted velocities from enu back to ecef" << std::endl;
    std::cout << "\t\tvX: " << xyz_vel[0] << "\tvY: " << xyz_vel[1] << "\tvZ: " << xyz_vel[2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service enu2xyz_vel");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert covariance from XYZ to ENU
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("xyz2enu_cov");

  double xyz_cov[3][3],enu_cov[3][3];

  xyz_cov[0][0] =  0.1527; xyz_cov[0][1] = -0.0244; xyz_cov[0][2] = 0.0267;
  xyz_cov[1][0] = -0.0244; xyz_cov[1][1] =  0.5772; xyz_cov[1][2] = -0.2432;
  xyz_cov[2][0] =  0.0267; xyz_cov[2][1] = -0.2432; xyz_cov[2][2] = 0.2687;

  std::cout << "\nConverting covariance" << std::endl;
  std::cout << "\tPxyz:\t" << xyz_cov[0][0] << "\t" << xyz_cov[0][1] << "\t" << xyz_cov[0][2] << std::endl;
  std::cout << "\t     \t" << xyz_cov[1][0] << "\t" << xyz_cov[1][1] << "\t" << xyz_cov[1][2] << std::endl;
  std::cout << "\t     \t" << xyz_cov[2][0] << "\t" << xyz_cov[2][1] << "\t" << xyz_cov[2][2] << std::endl;

  for(int i=0;i<3;i++)
    for(int j=0;j<3;j++)
      srv.request.xyz_cov[3*i+j] = xyz_cov[i][j];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    for(int i=0;i<3;i++)
      for(int j=0;j<3;j++)
        enu_cov[i][j]=srv.response.enu_cov[3*i+j];

    std::cout << "\n\tSuccessfully converted covariance from ecef to enu" << std::endl;
    std::cout << "\t\tPenu:\t" << enu_cov[0][0] << "\t" << enu_cov[0][1] << "\t" << enu_cov[0][2] << std::endl;
    std::cout << "\t\t     \t" << enu_cov[1][0] << "\t" << enu_cov[1][1] << "\t" << enu_cov[1][2] << std::endl;
    std::cout << "\t\t     \t" << enu_cov[2][0] << "\t" << enu_cov[2][1] << "\t" << enu_cov[2][2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service xyz2enu_cov");
    return 1;
  }

  ////////////////////////////////////////////////////////////////////////////////
  // Convert covariance from ENU to XYZ
  ////////////////////////////////////////////////////////////////////////////////
  client = nh.serviceClient<wgs_conversions::WgsConversion>("enu2xyz_cov");

  for(int i=0;i<3;i++)
    for(int j=0;j<3;j++)
      srv.request.enu_cov[3*i+j] = enu_cov[i][j];

  srv.request.ref_lla[0] = ref_lla[0];
  srv.request.ref_lla[1] = ref_lla[1];
  srv.request.ref_lla[2] = ref_lla[2];

  if (client.call(srv)){
    for(int i=0;i<3;i++)
      for(int j=0;j<3;j++)
        xyz_cov[i][j]=srv.response.xyz_cov[3*i+j];

    std::cout << "\n\tSuccessfully converted covariance from enu back to ecef" << std::endl;
    std::cout << "\t\tPxyz:\t" << xyz_cov[0][0] << "\t" << xyz_cov[0][1] << "\t" << xyz_cov[0][2] << std::endl;
    std::cout << "\t\t     \t" << xyz_cov[1][0] << "\t" << xyz_cov[1][1] << "\t" << xyz_cov[1][2] << std::endl;
    std::cout << "\t\t     \t" << xyz_cov[2][0] << "\t" << xyz_cov[2][1] << "\t" << xyz_cov[2][2] << std::endl;
  }else{
    ROS_ERROR("Failed to call service enu2xyz_cov");
    return 1;
  }

  return 0;
}
